File: src/main/java/me/kikugie/techutils/render/litematica/LitematicRenderManager.java
package me.kikugie.techutils.render.litematica;

import com.mojang.blaze3d.systems.RenderSystem;
import fi.dy.masa.litematica.gui.GuiSchematicBrowserBase;
import fi.dy.masa.malilib.config.IConfigOptionListEntry;
import fi.dy.masa.malilib.gui.widgets.WidgetFileBrowserBase;
import me.kikugie.techutils.config.Configs;
import net.minecraft.client.MinecraftClient;

import java.util.HashMap;
import java.util.Map;

public class LitematicRenderManager {
    private static LitematicRenderManager instance;
    private final Map<WidgetFileBrowserBase.DirectoryEntry, LitematicRenderer> rendererCache = new HashMap<>();
    private final GuiSchematicBrowserBase gui;
    private final int slant;
    private LitematicRenderer currentRenderer;
    private int renderX;
    private int renderY;
    private int viewportSize;
    private boolean validDragging = false;

    private LitematicRenderManager(GuiSchematicBrowserBase gui) {
        this.gui = gui;
        this.slant = Configs.LitematicConfigs.RENDER_SLANT.getIntegerValue();
    }

    public static LitematicRenderManager init(GuiSchematicBrowserBase gui) {
        instance = new LitematicRenderManager(gui);
        return instance;
    }

    public static void reset() {
        instance = null;
    }

    public static LitematicRenderManager getInstance() {
        return instance;
    }

    public void setCurrentRenderer(WidgetFileBrowserBase.DirectoryEntry entry) {
        if (this.gui == null) return;
        if (rendererCache.containsKey(entry)) {
            this.currentRenderer = rendererCache.get(entry);
        } else {
            var renderer = new LitematicRenderer(entry, this.gui, this.slant);
            rendererCache.put(entry, renderer);
            this.currentRenderer = renderer;
        }
    }

    public void renderCurrent(int x, int y, int viewportSize) {
        if (currentRenderer == null) return;
        this.renderX = x;
        this.renderY = y;
        this.viewportSize = viewportSize;

        if (Configs.LitematicConfigs.RENDER_ROTATION_MODE.getStringValue().equals(RotationMode.MOUSE_POS.mode)) {
            double mouseX = MinecraftClient.getInstance().mouse.getX();
            int windowWidth = MinecraftClient.getInstance().getWindow().getFramebufferWidth();
            currentRenderer.angle = (int) (mouseX / windowWidth * Configs.LitematicConfigs.ROTATION_FACTOR.getDoubleValue() * 360);
        } else if (Configs.LitematicConfigs.RENDER_ROTATION_MODE.getStringValue().equals(RotationMode.FREE_SPIN.mode)) {
            long tick = MinecraftClient.getInstance().world.getTime();
            currentRenderer.angle = (int) (tick * Configs.LitematicConfigs.ROTATION_FACTOR.getDoubleValue() * 2 % 360);
        }
        this.currentRenderer.render(RenderSystem.getModelViewStack(), x, y, viewportSize);

    }

    public void mouseScrolled(double mouseX, double mouseY, double amount) {
        if (!Configs.LitematicConfigs.RENDER_ROTATION_MODE.getStringValue().equals(RotationMode.SCROLL.mode) || !isInViewPort(mouseX, mouseY))
            return;
        currentRenderer.angle += amount * Configs.LitematicConfigs.ROTATION_FACTOR.getDoubleValue() * 10;
    }

    public void mouseDragged(double deltaX) {
        if (!Configs.LitematicConfigs.RENDER_ROTATION_MODE.getStringValue().equals(RotationMode.DRAG.mode) || !validDragging)
            return;
        currentRenderer.angle += deltaX * Configs.LitematicConfigs.ROTATION_FACTOR.getDoubleValue();
    }

    public void mouseReleased() {
        this.validDragging = false;
    }

    public void mouseClicked(double mouseX, double mouseY, int mouseButton) {
        if (isInViewPort(mouseX, mouseY) && mouseButton == 0) {
            this.validDragging = true;
        }
    }

    private boolean isInViewPort(double mouseX, double mouseY) {
        return mouseX > renderX && mouseY > renderY && mouseX < renderX + viewportSize && mouseY < renderY + viewportSize;
    }

    public enum RotationMode implements IConfigOptionListEntry {
        MOUSE_POS("Mouse position"),
        DRAG("Drag"),
        SCROLL("Scroll"),
        FREE_SPIN("Free spin");

        private final String mode;

        RotationMode(String mode) {
            this.mode = mode;
        }

        @Override
        public String getStringValue() {
            return mode;
        }

        @Override
        public String getDisplayName() {
            return mode;
        }

        @Override
        public IConfigOptionListEntry cycle(boolean forward) {
            int mod = forward ? 1 : -1;
            return values()[(this.ordinal() + mod) % values().length];
        }

        @Override
        public IConfigOptionListEntry fromString(String value) {
            RotationMode[] vals = values();

            for (RotationMode temp : vals) {
                if (temp.mode.equalsIgnoreCase(value)) {
                    return temp;
                }
            }

            return RotationMode.DRAG;
        }
    }
}

File: src/main/kotlin/me/kikugie/techutils/render/litematica/LitematicRenderManager.kt
package me.kikugie.techutils.render.litematica

import com.mojang.blaze3d.systems.RenderSystem
import fi.dy.masa.litematica.gui.GuiSchematicBrowserBase
import fi.dy.masa.malilib.config.IConfigOptionListEntry
import fi.dy.masa.malilib.gui.widgets.WidgetFileBrowserBase
import me.kikugie.techutils.config.Configs
import net.minecraft.client.MinecraftClient

class LitematicRenderManager private constructor(gui: GuiSchematicBrowserBase) {
    private val rendererCache: MutableMap<WidgetFileBrowserBase.DirectoryEntry, LitematicRenderer> =
        HashMap()
    private val gui: GuiSchematicBrowserBase?
    private val slant: Int
    private var currentRenderer: LitematicRenderer? = null
    private var renderX = 0
    private var renderY = 0
    private var viewportSize = 0
    private var validDragging = false

    init {
        this.gui = gui
        slant = Configs.LitematicConfigs.RENDER_SLANT.integerValue
    }

    fun setCurrentRenderer(entry: WidgetFileBrowserBase.DirectoryEntry?) {
        if (gui == null) return
        if (rendererCache.containsKey(entry)) {
            currentRenderer = rendererCache[entry]
        } else {
            val renderer = LitematicRenderer(entry!!, gui, slant)
            rendererCache[entry] = renderer
            currentRenderer = renderer
        }
    }

    fun renderCurrent(x: Int, y: Int, viewportSize: Int) {
        if (currentRenderer == null) return
        renderX = x
        renderY = y
        this.viewportSize = viewportSize
        if (Configs.LitematicConfigs.RENDER_ROTATION_MODE.stringValue.equals(RotationMode.FREE_SPIN.displayName)) {
            val mouseX = MinecraftClient.getInstance().mouse.x
            val windowWidth = MinecraftClient.getInstance().window.framebufferWidth
            currentRenderer!!.angle = (mouseX / windowWidth * 360).toInt()
        }
        currentRenderer!!.render(RenderSystem.getModelViewStack(), x, y, viewportSize)
    }

    fun mouseScrolled(mouseX: Double, mouseY: Double, amount: Double) {
        if (!Configs.LitematicConfigs.RENDER_ROTATION_MODE.stringValue
                .equals(RotationMode.SCROLL.displayName) || !isInViewPort(mouseX, mouseY)
        ) return
        currentRenderer!!.angle += (amount * 10).toInt()
    }

    fun mouseDragged(deltaX: Double) {
        if (!Configs.LitematicConfigs.RENDER_ROTATION_MODE.stringValue
                .equals(RotationMode.DRAG.displayName) || !validDragging
        ) return
        currentRenderer!!.angle += (deltaX * 2).toInt()
    }

    fun mouseReleased() {
        validDragging = false
    }

    fun mouseClicked(mouseX: Double, mouseY: Double, mouseButton: Int) {
        if (isInViewPort(mouseX, mouseY) && mouseButton == 0) {
            validDragging = true
        }
    }

    private fun isInViewPort(mouseX: Double, mouseY: Double): Boolean {
        return mouseX > renderX && mouseY > renderY && mouseX < renderX + viewportSize && mouseY < renderY + viewportSize
    }

    enum class RotationMode(private val displayName: String) : IConfigOptionListEntry {
        FREE_SPIN("Free Spin"), DRAG("Drag"), SCROLL("Scroll");

        override fun getStringValue(): String {
            return displayName
        }

        override fun getDisplayName(): String {
            return displayName
        }

        override fun cycle(forward: Boolean): IConfigOptionListEntry {
            val mod = if (forward) 1 else -1
            return values()[(ordinal + mod) % values().size]
        }

        override fun fromString(value: String): IConfigOptionListEntry {
            val vals = values()
            for (temp in vals) {
                if (temp.displayName.equals(value, ignoreCase = true)) {
                    return temp
                }
            }
            return DRAG
        }
    }

    companion object {
        @JvmStatic
        var instance: LitematicRenderManager? = null
            private set

        fun init(gui: GuiSchematicBrowserBase): LitematicRenderManager? {
            instance = LitematicRenderManager(gui)
            return instance
        }

        @JvmStatic
        fun reset() {
            instance = null
        }
    }
}